fussy.el

```diff
@@ -1082,13 +1082,27 @@ highlighting."
       (when (fboundp 'fzf-native-make-default-slab)
         (setf fussy--fzf-native-slab (fzf-native-make-default-slab)))))
 
+(defun fussy--fzf-native-fix-score-indices (str score)
+  "Fix SCORE indices of `fzf-native' if STR is multibyte."
+  (if (or (not (multibyte-string-p str)) (null score))
+      score
+    ;; fzf-native makes score indices as byte position.
+    ;; But we want it as character position.
+    (let ((idx (cl-loop for c across str
+                        for i = 0 then (1+ i)
+                        vconcat (make-vector (string-bytes (char-to-string c)) i))))
+      (cons (car score) (mapcar (lambda (x) (aref idx x)) (cdr score))))))
+
+
 (defun fussy-fzf-native-score (str query &rest _args)
   "Score STR for QUERY using `fzf-native'."
   (require 'fzf-native)
   (when (fboundp 'fzf-native-score)
     (let ((str (funcall fussy-remove-bad-char-fn str))
           (query (fussy-encode-coding-string query)))
-      (fzf-native-score str query (fussy--fzf-native-slab)))))
+      (fussy--fzf-native-fix-score-indices
```

Comment:
Wonder if it's possible for fzf-native to check this instead. @dangduc

Reply:
I think an advice is better here. @buzztaiki I'd accept a PR to [fzf-native](https://github.com/dangduc/fzf-native).

```emacs-lisp
(defun fzf-native--fix-score-indices (fzf-native-score-fn &rest args)
  "Fix SCORE indices of `fzf-native-score' if STR is multibyte."
  (let ((res (apply fzf-native-score-fn args))
        (str (car args)))
    (if (or (not (multibyte-string-p str)) (null res))
        res
      ;; fzf-native makes score indices as byte position.
      ;; But we want it as character position.
      (let ((idx (cl-loop for c across str
                          for i = 0 then (1+ i)
                          vconcat (make-vector (string-bytes (char-to-string c)) i))))
        (cons (car res) (mapcar (lambda (x) (aref idx x)) (cdr res)))))))

(fzf-native-score "ãã±ã¢ã³.txt" "txt")
; => (80 25 26 27)

(advice-add 'fzf-native-score :around #'fzf-native--fix-score-indices)

(fzf-native-score "ãã±ã¢ã³.txt" "txt")
; => (80 13 14 15)
```

Although this PR fixes indices, the actual scoring is unaddressed. Due to character distance being a factor in scoring. And distance assumes single byte chars.

```emacs-lisp
(fzf-native-score "aaaatxt" "txt")
; => (56 4 5 6)

(fzf-native-score "asdftxt" "txt")
; => (56 4 5 6)

(fzf-native-score "txt" "txt")
; => (80 4 5 6)
```

Reply:
@dangduc Which is better to use the advice or to call `fzf-native--fix-score-indices` in ` fzf_native_score`?

Reply:
I would take a PR to fzf-native that adds the advice function `fzf-native--fix-score-indices`. Then all you'd need to do is to configure the advice. For example if using straight.el:
```diff
(use-package fzf-native
  :ensure t
  :straight
  (fzf-native
   :repo "dangduc/fzf-native"
   :host github
   :files (:defaults "bin"))
  :config
+ (advice-add 'fzf-native-score :around #'fzf-native--fix-score-indices)
  (setq fussy-score-fn 'fussy-fzf-native-score)
 (fzf-native-load-dyn))
 ```

Alternatively the PR could just enable the advice by default in fzf-native.el package. So you wouldn't need to do `advice-add` in package declaration.

```diff
; fzf-native.el file
+(defun fzf-native--fix-score-indices (fzf-native-score-fn &rest args)
+  "Fix SCORE indices of `fzf-native-score' if STR is multibyte."
+  (let ((res (apply fzf-native-score-fn args))
+        (str (car args)))
+    (if (or (not (multibyte-string-p str)) (null res))
+        res
+      ;; fzf-native makes score indices as byte position.
+      ;; But we want it as character position.
+      (let ((idx (cl-loop for c across str
+                          for i = 0 then (1+ i)
+                          vconcat (make-vector (string-bytes (char-to-string c)) i))))
+        (cons (car res) (mapcar (lambda (x) (aref idx x)) (cdr res)))))))
+
+ (advice-add 'fzf-native-score :around #'fzf-native--fix-score-indices)
```

An alternative approach to advising is have the dynamic module code call `fzf-native--fix-score-indices`. But I think we want to limit calls to the emacs process from the dynamic module.

Reply:
In increasing desirability, I'd want the fix to be:

1. Elisp code that fixes the indices, e.g. @buzztaiki's original PR or my suggestion to fzf-native.el.
2. C code that fixes the indices.
3. fzf.c fixes the indices.
4. fzf.c fixes the scoring + indices.

Reply:
Thanks, I will enable the advice by default.

Reply:
I've send PR in https://github.com/dangduc/fzf-native/pull/15